### PR TITLE
feat: Make the whole thing a Python wheel

### PR DESCRIPTION
**Pull request by lukapeschke**
This "works" (quick and dirty), but perf is okay-ish (consistently below 10s on my machine with our fat excel).

One thing I don't like is the memory usage though:
image

A few notes on this:
* Linear increase is almost 100% the loop over all sheets + `RecordBatch` instantiation
* Highest spike just after is probably the double vector allocation + transformation to `PyBytes`. The straight drop right after would confirm this, since Rust would destroy all unused vecs when yielding to Python
* I suspect the last spike to be the `bytes -> RecordBatch -> DataFrame`. I'm afraid there's no fast solution (doesn't mean there's no solution at all though, see below)

My next steps:
* Make everything lazy:
    * First, implement python's iterator protocol in rust to return an iterator over record batches rather than a vec (useless for our fat csv, useful when iterating over multiple sheets)
    * Later on, maybe build the arrow schema first, and then construct row objects rather than columns. That way, we could yield record batches that would be chunks of a sheet (useful for preview in a single file). And the dataset insertion in our DB is also chunked, so that would work
* If I'm bored and/or frustrated by the vector spike, instantiate `PyBytes` objects with `unsafe` to prevent the double alloc
* Lastly, but that might take longer , PyO3 provides rust bindings for numpy: https://github.com/PyO3/rust-numpy . We could directly provide numpy objects to the python layer, and build pandas `DataFrame` objects from those (and hope really hard that pandas is smart with allocs when a df is built from numpy arrays)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+mod core;
+
+use anyhow::{Context, Result};
+use pyo3::{prelude::*, types::PyBytes};
+
+use crate::core::{extract_sheets, record_batch_to_bytes};
+
+/// Formats the sum of two numbers as string.
```
Comment (PrettyWood) [inline src/lib.rs:8]:
```suggestion
/// Formats the sum of two numbers as string.
```
huhu reste du template maturin ça ;D

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#[pyfunction]
+fn read_excel<'p>(py: Python<'p>, path: &str) -> Result<Vec<&'p PyBytes>> {
+    // FIXME: Allocating two vecs here, extract_sheets should return an Iterator
+    let sheets = extract_sheets(path).with_context(|| format!("could not load file at {path}"))?;
+    sheets
+        .iter()
+        // FIXME: Allocating AGAIN here (PyBytes::new clones, we'll probably have to use some unsafe
+        // shit here, such as from_raw_ptr)
```
Comment (PrettyWood) [inline src/lib.rs:16]: Chaud pour une ptite session pour en parler !

**Review by PrettyWood — COMMENTED** [review]
Top en premier jet !

**Comment by jGundermann** [comment]
Nice work @lukapeschke! Pour rappel ce xslx que je t'ai donner s'ouvre en plus de deux minutes avec libre-office ou avec ToucanToco actuellement. Donc si on arrive a faire ça en 10 secondes on a un gain de perf énorme.

**Review by PrettyWood — CHANGES_REQUESTED** [review]
Ce serait bien aussi de rajouter un `.pyi` et un `py.typed` pour l'intellisense en python

File: src/lib.rs (hunk 1 above)
Comment (lukapeschke) [reply src/lib.rs]: oops :grimacing:

**Comment by lukapeschke** [comment]
For reference, a few additional benchamrks:

Memory usage with `pandas` + `openpyxl` (`pd.read_excel('fat_excel.xlsx')`). This takes ~1min40s on my machine:
image

In-memory size of DataFrame according to pandas (`df.info(memory_usage="deep")`): 
```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277211 entries, 0 to 277210
Data columns (total 41 columns):
 #   Column                                            Non-Null Count   Dtype
---  ------                                            --------------   -----
 0   BU                                                275944 non-null  object
 1   SubBU                                             276292 non-null  object
 2   Ou Name                                           277211 non-null  object
 3   Site                                              276575 non-null  object
 4   Site Designation                                  276213 non-null  object
 5   Oracle Item Mass Allocation                       273956 non-null  object
 6   Oracle Item Mass Allocation Description           272746 non-null  object
 7   Oracle Item Line                                  229826 non-null  object
 8   Oracle Item Line Description                      229826 non-null  object
 9   Oracle Item Line_1                                229826 non-null  object
 10  Oracle Item Line Description_1                    229826 non-null  object
 11  Item Sales Account                                277183 non-null  object
 12  Item BU GL Code                                   277183 non-null  object
 13  Item Site GL Code                                 277183 non-null  object
 14  Item Departement GL Code                          277183 non-null  object
 15  Item Account GL Code                              277183 non-null  object
 16  Item Product Group GL Code                        277183 non-null  object
 17  Country Name - Bill-to                            277209 non-null  object
 18  Intercompany Type Code - Bill-to                  277172 non-null  object
 19  Account Number - Bill-to                          277209 non-null  object
 20  Party Name - Bill-to                              277209 non-null  object
 21  Country Name - Bill-to_1                          277209 non-null  object
 22  EIA Sales Person Number - Orders                  269695 non-null  object
 23  EIA Sales Person Name - Orders                    269695 non-null  object
 24  EIA Sales Person Branch - Orders                  0 non-null       object
 25  LS Conso Perimeter - Bill-to                      277172 non-null  object
 26  Account GL Code - Bill-to                         277202 non-null  object
 27  Account GL Description - Bill-to                  277202 non-null  object
 28  Intercompany Type Code - Bill-to_1                277172 non-null  object
 29  Intercompany Type Description - Bill-to           277172 non-null  object
 30  Invoice Fiscal Month                              277211 non-null  float64
 31  Invoice Fiscal Year                               277211 non-null  float64
 32  Invoiced Quantity                                 277211 non-null  float64
 33  Charge Other Cost (filter Order Currency)         0 non-null       object
 34  Charge LS DE Admin (filter Order Currency)        0 non-null       object
 35  Charge DE Shipment (filter Order Currency)        0 non-null       object
 36  Charge Freight Manual (filter Order Currency)     230666 non-null  float64
 37  Charge Freight (filter Order Currency)            221070 non-null  float64
 38  Charge Metal (filter Order Currency)              237856 non-null  float64
 39  Charge Energy (filter Order Currency)             0 non-null       object
 40  Invoiced Selling Price (filter Invoice Currency)  277211 non-null  float64
dtypes: float64(7), object(34)
memory usage: 545.6 MB
```
So we can consider this as a hard lower limit when it comes to non-lazy loading.

For reference, the size of the `.xlsx` file is 49M, and the size of the DataFrame dumped as CSV is 91M

File: src/core.rs [hunk 3]
```diff
@@ -53,10 +53,7 @@ pub fn extract_sheets(path: &str) -> Result<Vec<RecordBatch>> {
                 datatypes::DataType::Float64 => create_float_array(&data, col, height),
                 datatypes::DataType::Utf8 => create_string_array(&data, col, height),
                 datatypes::DataType::Null => Arc::new(NullArray::new(height - 1)),
-                d => {
-                    println!("{:?}", d);
-                    todo!();
-                }
+                _ => panic!("Unreachable code"),
```
Comment (PrettyWood) [inline src/core.rs:56]:
```suggestion
                _ => unreachable!(),
```

**Review by PrettyWood — APPROVED** [review]

**Pull request merged**